File: backend/app/models/milestone.py
from datetime import datetime
from enum import Enum

from sqlalchemy import (
    Column,
    DateTime,
    Enum as SqlEnum,
    ForeignKey,
    Integer,
    String,
    Text,
)

from app.db.database import Base


class MilestoneStatus(str, Enum):
    OPEN = "OPEN"
    CLOSED = "CLOSED"


class Milestone(Base):
    __tablename__ = "milestones"

    id = Column(
        Integer,
        primary_key=True,
        index=True,
    )

    project_id = Column(
        Integer,
        ForeignKey(
            "projects.id",
            ondelete="CASCADE",
        ),
        nullable=False,
    )

    title = Column(
        String,
        nullable=False,
    )

    description = Column(
        Text,
        nullable=True,
    )

    status = Column(
        SqlEnum(MilestoneStatus),
        default=MilestoneStatus.OPEN,
        nullable=False,
    )

    due_date = Column(
        DateTime,
        nullable=True,
    )

    created_at = Column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
    )

    updated_at = Column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )

File: frontend/src/types/milestone.ts
// Milestone Status
export type MilestoneStatus =
  | "OPEN"
  | "CLOSED";


// GET milestone
export interface Milestone {
  id: number;
  project_id: number;
  title: string;
  description: string | null;
  status: MilestoneStatus;
  due_date: string | null;
  created_at: string;
  updated_at: string;
}


// POST milestone
export interface CreateMilestoneRequest {
  title: string;
  description?: string;
  due_date?: string | null;
}


// PATCH milestone
export interface UpdateMilestoneRequest {
  title?: string;
  description?: string;
  status?: MilestoneStatus;
  due_date?: string | null;
}